test/integration/federation/framework/crud.go

```diff
@@ -0,0 +1,209 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	pkgruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
+)
+
+// CRUDHelper exercises Create/Read/Update/Delete operations for
+// federated types via the Federation API and validates that the
+// results of those operations are propagated to clusters that are
+// members of a federation.
+type CRUDHelper struct {
+	tl             TestLogger
+	adapter        ResourceAdapter
+	kind           string
+	clusterClients []clientset.Interface
+	waitInterval   time.Duration
+	// Federation operations will use wait.ForeverTestTimeout.  Any
+	// operation that involves member clusters may take longer due to
+	// propagation latency.
+	clusterWaitTimeout time.Duration
+}
+
+func NewCRUDHelperWithTimeout(testLogger TestLogger, adapter ResourceAdapter, clusterClients []clientset.Interface, waitInterval, clusterWaitTimeout time.Duration) *CRUDHelper {
+	return &CRUDHelper{
+		tl:                 testLogger,
+		adapter:            adapter,
+		kind:               adapter.Kind(),
+		clusterClients:     clusterClients,
+		waitInterval:       waitInterval,
+		clusterWaitTimeout: clusterWaitTimeout,
+	}
+}
+
+func NewCRUDHelper(t *testing.T, adapter ResourceAdapter, clusterClients []clientset.Interface) *CRUDHelper {
+	logger := &IntegrationLogger{t}
+	return NewCRUDHelperWithTimeout(logger, adapter, clusterClients, DefaultWaitInterval, wait.ForeverTestTimeout)
+}
+
+func (c *CRUDHelper) CheckLifecycle(obj pkgruntime.Object) {
+	updatedObj := c.CheckCreate(obj)
+	c.CheckUpdate(updatedObj)
+
+	// Validate the golden path - removal of dependents
+	orphanDependents := false
+	c.CheckDelete(updatedObj, &orphanDependents)
+}
+
+func (c *CRUDHelper) CheckCreate(obj pkgruntime.Object) pkgruntime.Object {
+	namespacedName := c.adapter.NamespacedName(obj)
+
+	c.tl.Logf("Creating federated %s %q", c.kind, namespacedName)
+
+	obj, err := c.adapter.FedCreate(obj)
+	if err != nil {
+		c.tl.Fatalf("Error creating federated %s %q : %v", c.kind, namespacedName, err)
+	}
+
+	c.checkPropagation(obj)
+
+	return obj
+}
+
+func (c *CRUDHelper) CheckUpdate(obj pkgruntime.Object) {
+	namespacedName := c.adapter.NamespacedName(obj)
+
+	// Insure that the provided object does not have labels, which
```

Comment:
Is this a sustainable strategy? It seems like any of the objects could have labels for valid reasons.

Would it be safer to use annotations?